src/features/events/components/LocationModal/index.tsx

```diff
@@ -1,42 +1,153 @@
+import LocationDetailsCard from './LocationDetailsCard';
+import LocationSearch from './LocationSearch';
 import messageIds from 'features/events/l10n/messageIds';
 import { useMessages } from 'core/i18n';
 import { ZetkinLocation } from 'utils/types/zetkin';
 
+interface StyleProps {
+  isLong: boolean;
+}
+
+const useStyles = makeStyles<Theme, StyleProps>(() => ({
+  overlay: {
+    bottom: ({ isLong }) => (isLong ? 64 : ''),
+    display: 'flex',
+    justifyContent: 'flex-end',
+    justifySelf: 'flex-end',
+    margin: 2,
+    position: 'absolute',
+    right: 32,
+    top: 32,
+    width: '30%',
+    zIndex: 1000,
+  },
+}));
+
+export type PendingLocation = {
+  lat: number;
+  lng: number;
+};
+
 interface LocationModalProps {
   locations: ZetkinLocation[];
+  onCreateLocation: (newLocation: Partial<ZetkinLocation>) => void;
   onMapClose: () => void;
   onSelectLocation: (location: ZetkinLocation) => void;
   open: boolean;
   locationId?: number;
 }
 
+const Map = dynamic(() => import('./Map'), { ssr: false });
 const LocationModal: FC<LocationModalProps> = ({
   locations,
+  onCreateLocation,
   onMapClose,
   onSelectLocation,
   open,
-  locationId,
+  locationId = null,
 }) => {
-  const Map = dynamic(() => import('./Map'), { ssr: false });
   const messages = useMessages(messageIds);
+  const [searchString, setSearchString] = useState('');
+  const [selectedLocationId, setSelectedLocationId] = useState(locationId);
+  const [focusedMarker, setFocusedMarker] = useState<
+    { lat: number; lng: number } | undefined
+  >();
+  const [pendingLocation, setPendingLocation] = useState<Pick<
+    ZetkinLocation,
+    'lat' | 'lng'
+  > | null>(null);
+
+  const selectedLocation = locations.find(
+    (location) => location.id === selectedLocationId
+  );
+
+  const isLong = !!pendingLocation || !!selectedLocation;
+  const classes = useStyles({ isLong });
+
+  useEffect(() => {
+    setSelectedLocationId(locationId);
+    setPendingLocation(null);
+  }, [open]);
 
   return (
     <Dialog fullWidth maxWidth="lg" onClose={onMapClose} open={open}>
       <Box padding={2}>
         <Map
-          locationId={locationId}
+          focusedMarker={focusedMarker}
           locations={locations}
-          onMapClose={onMapClose}
-          onSelectLocation={(location: ZetkinLocation) =>
-            onSelectLocation(location)
-          }
+          onMapClick={(latlng: PendingLocation) => {
+            setSelectedLocationId(null);
+            setPendingLocation(latlng);
+          }}
+          onMarkerClick={(locationId: number) => {
+            const location = locations.find(
+              (location) => location.id === locationId
+            );
+            if (!location?.lat || !location?.lng) {
+              return;
+            }
+            setPendingLocation(null);
+            setSelectedLocationId(location.id);
+            setFocusedMarker({ lat: location.lat, lng: location.lng });
```

Comment:
Why is this separate from the `selectedLocation` state that already exists separately?

Reply:
honestly do not remember, will refactor